Doc update

outoftardis opened the pull request:
# Description

- Added [CPU vs. GPU topic](https://outoftardis.github.io/daal/cpu-vs-gpu.html) + provided notes about the differences across affected topics
- Added info about `hist` method for decision forest([description](https://outoftardis.github.io/daal/dev_guide/algorithms/decision_forest/decision-forest.html#inexact-computation-method), [regression](https://outoftardis.github.io/daal/dev_guide/algorithms/decision_forest/decision-forest-regression.html#training), [classification](https://outoftardis.github.io/daal/dev_guide/algorithms/decision_forest/decision-forest-classification.html#training)), `brute force` method for [kNN](https://outoftardis.github.io/daal/dev_guide/algorithms/k_nearest_neighbors/k-nearest-neighbors-knn-classifier.html), and `thunder` method for [SVM](https://outoftardis.github.io/daal/dev_guide/algorithms/svm/support-vector-machine-classifier.html)
- Updated [KMeans](https://outoftardis.github.io/daal/dev_guide/algorithms/kmeans/k-means-clustering.html)

output here: https://outoftardis.github.io/daal/index.html

PetrovKP commented on docs/source/dev_guide/algorithms/svm/support-vector-machine-classifier.rst:
```diff
@@ -201,31 +203,26 @@ Training and Prediction.
+     - The number of classes.
+   * - ``kernel``
+     - Pointer to object of the ``KernelIface`` class
+     - The kernel function. By default, the algorithm uses a linear kernel.
```
Please, update examples:

```
  .. tab:: DPC++
   Batch Processing:
   -  :ref:`svm_two_class_dense_batch.cpp>`

  .. tab:: C++
   Batch Processing:
   -  :cpp_example:`svm_two_class_boser_dense_batch.cpp <svm/svm_two_class_boser_dense_batch.cpp>`
   -  :cpp_example:`svm_two_class_boser_csr_batch.cpp <svm/svm_two_class_boser_csr_batch.cpp>`
   -  :cpp_example:`svm_two_class_thunder_dense_batch.cpp <svm/svm_two_class_thunder_dense_batch.cpp>`
   -  :cpp_example:`svm_two_class_thunder_csr_batch.cpp <svm/svm_two_class_thunder_csr_batch.cpp>`

.. tab:: Java*
   Batch Processing:
   -  :java_example:`SVMTwoClassBoserDenseBatch.java <svm/SVMTwoClasBosersDenseBatch.java>`
   -  :java_example:`SVMTwoClassBoserCSRBatch.java <svm/SVMTwoClassBoserCSRBatch.java>`
   -  :java_example:`SVMTwoClassThunderDenseBatch.java <svm/SVMTwoClassThunderDenseBatch.java>`
   -  :java_example:`SVMTwoClassThunderCSRBatch.java <svm/SVMTwoClassThunderCSRBatch.java>`
```

outoftardis replied:
fixed, thanks

amgrigoriev commented on docs/source/dev_guide/algorithms/dbscan/distributed-steps/step-6.rst:
```diff
@@ -50,6 +50,9 @@ In this step, the DBSCAN algorithm has the following parameters:
        which in worst case can be :math:`O(|\text{number of observations}|^2)`. However, in general, performance may be better.
 
+
+       .. note:: ``memorySavingMode`` could be ``false`` only on CPU.
```
There is no implementation for distributed dbscan steps on GPU

outoftardis replied:
fixed

amgrigoriev commented on docs/source/dev_guide/algorithms/dbscan/computation-batch.rst:
```diff
@@ -53,6 +53,9 @@ The DBSCAN clustering algorithm has the following parameters:
        which in worst case can be :math:`O(|\text{number of observations}|^2)`. However, in general, performance may be better.
+
+       .. note:: ``memorySavingMode`` could be ``false`` only on CPU.
+
```
For current implementation, weights are ignored on GPU. Support for weights will be added later

outoftardis replied:
fixed

amgrigoriev commented on docs/source/dev_guide/algorithms/k_nearest_neighbors/k-nearest-neighbors-knn-classifier.rst:
```diff
@@ -56,10 +53,23 @@ Given a positive integer parameter :math:`k` and a test observation
-non-leaf node of a tree contains the identifier of a feature along
+The library provides kNN classification with the following methods:
+
+- K-D tree
```
K-D tree method is only supported on CPU

outoftardis replied:
fixed

outoftardis commented on docs/source/bibliography.rst:
```diff
@@ -185,6 +185,12 @@ For more information about algorithms implemented in |short_name|, refer to the
    Co., Inc. Boston, MA, USA, 2005, ISBN: 032132136.
 
+.. [Wen2018]
+
```
fixed

outoftardis commented on docs/source/dev_guide/algorithms/decision_forest/decision-forest.rst:
```diff
@@ -86,6 +100,9 @@ each terminal node :math:`t`:
    :math:`\Delta i` in all :math:`s_{j,t}` splits.
 -  Apply this procedure recursively to :math:`t_L` and :math:`t_R`.
 
+Creation of a Bootstrap Set
```
fixed

outoftardis commented on docs/source/dev_guide/algorithms/decision_forest/decision-forest.rst:
```diff
@@ -49,10 +49,22 @@ internal nodes :math:`t` represent a binary test (split) dividing their
+Inexact Computation Method
+++++++++++++++++++++++++++
+
+In inexact method, also called histogram method, only a selected subset of splits is considered for computation of a best split.
```
rephrased

outoftardis commented on docs/source/dev_guide/algorithms/svm/support-vector-machine-classifier.rst:
```diff
@@ -139,59 +143,57 @@ Training and Prediction.
+
+       For CPU:
+
+        - ``defaultDense`` – Boser method [Boser92]_
```
I'll add this closer to beta08, since right now it's not supported yet

PetrovKP replied:
Ok

outoftardis commented on docs/source/dev_guide/algorithms/svm/support-vector-machine-classifier.rst:
```diff
@@ -139,59 +143,57 @@ Training and Prediction.
+     - ``true``
+     - A flag that enables use of a shrinking optimization technique.
+
+       .. note:: This parameter is only supported for ``defaultDense`` method.
```
There is more variety there. I'd argue that one small difference between two method doesn't justify creation of a new column. If you disagree, I'll add it, of course.

PetrovKP replied:
I think it should be consistent. And in the future, specific parameters for each method may be added. But you can leave it now. :)

outoftardis commented on docs/source/dev_guide/algorithms/k_nearest_neighbors/k-nearest-neighbors-knn-classifier.rst:
```diff
@@ -34,16 +36,11 @@ multidimensional feature space.
+The problem is to build a kNN classifier given the following:
 
-Given n feature vectors :math:`x_1 = (x_{11}, \ldots, x_{1p}), \ldots, x_n = (x_{n1}, \ldots, x_{np})` of size :math:`p` and a vector
-of class labels :math:`y = (y_1, \ldots, y_n)`, where :math:`y_i \in \{0, 1, \ldots, C-1\}`
-and :math:`C` is the number of classes, describes the
-class to which the feature vector :math:`x_i` belongs, the problem is
-to build a kNN classifier.
+- :math:`n` feature vectors :math:`x_1 = (x_{11}, \ldots, x_{1p}), \ldots, x_n = (x_{n1}, \ldots, x_{np})` of size :math:`p` 
+- a vector of class labels :math:`y = (y_1, \ldots, y_n)`, there :math:`C` is the number of classes and 
+  :math:`y_i \in \{0, 1, \ldots, C-1\}` describes the class the feature vector :math:`x_i` belongs to
```
It was done to increase readability, it's essentially the same text.

outoftardis replied:
ok, rolled back those changes

outoftardis commented on docs/source/dev_guide/algorithms/k_nearest_neighbors/k-nearest-neighbors-knn-classifier.rst:
```diff
@@ -68,9 +78,18 @@ Feature vectors from any bucket belong to the region of the space
+Brute Force
+-----------
+
+Brute Force is a naïve neighbor search that computes distances between all pairs of feature vactors
```
rephrased

outoftardis commented on docs/source/dev_guide/algorithms/k_nearest_neighbors/k-nearest-neighbors-knn-classifier.rst:
```diff
@@ -99,12 +119,23 @@ threshold. Once the threshold is reached, PANDA stops growing the
+Training using Brute Force
+++++++++++++++++++++++++++
+
+For training with the Brute Force method, the algorithm uses
```
rephrased

averbukh commented on docs/source/dev_guide/algorithms/k_nearest_neighbors/k-nearest-neighbors-knn-classifier.rst:
```diff
@@ -125,7 +154,7 @@ Training
 For a description of the input and output, refer to Usage Model: Training and Prediction.
 
-At the training stage, K-D tree based kNN classifier has the
+At the training stage, both Brute Force and K-D tree based kNN classifiesr have the
```
Misspelling: 'classifiesr'

outoftardis replied:
fixed

averbukh commented on docs/source/dev_guide/algorithms/k_nearest_neighbors/k-nearest-neighbors-knn-classifier.rst:
```diff
@@ -170,7 +199,7 @@ Prediction
 For a description of the input and output, refer to Usage Model: Training and Prediction.
 
-At the prediction stage, K-D tree based kNN classifier has the
+At the prediction stage, both Brute Force and K-D tree based kNN classifiesr have the
```
Misspelling.

outoftardis replied:
fixed

averbukh commented on docs/source/dev_guide/algorithms/kmeans/includes/computation-batch.rst:
```diff
@@ -41,39 +43,58 @@ The K-Means clustering algorithm has the following parameters:
+     - ``true``
      - A flag that enables computation of assignments, that is, assigning cluster indices to respective observations.
+   * - ``resultsToEvaluate``
+     - ``computeCentroids`` | ``computeAssignments`` | ``computeExactObjectiveFunction``
```
Does not look consistent with other algorithms

outoftardis replied:
that's the input that was provided to me, and it was explained that it should be written like this

averbukh replied:
OK.  I discussed it offline with Kirill. Let's leave it as it is for now.

averbukh commented on docs/source/dev_guide/algorithms/kmeans/includes/computation-batch.rst:
```diff
@@ -41,39 +43,58 @@ The K-Means clustering algorithm has the following parameters:
+     - ``computeCentroids`` | ``computeAssignments`` | ``computeExactObjectiveFunction``
+     - The 64-bit integer flag that specifies which extra characteristics of the K-Means to compute.
+     
+       Provide one of the following values to request a single characteristic or use bitwise OR to request a combination of the characteristics:
+       
+       - ``computeCentroids`` for computation centroids.
+       - ``computeAssignments`` for computation of assignments, that is, assigning cluster indices to respective observations.
+       - ``computeExactObjectiveFunction`` for computation of exact ObjectiveFunction.
```
It's breaks previous logic

outoftardis replied:
not sure I understand what you mean

averbukh replied:
I discussed it offline with Kirill. Let's leave it as it is for now.

averbukh commented on docs/source/dev_guide/algorithms/logistic_regression/logistic-regression.rst:
```diff
@@ -166,17 +166,20 @@ parameters:
      - :math:`0`
      - L1 regularization coefficient
+
+       .. note:: This parameter is not supported on GPU.
```
Very unclear. It should affect linkage with selected solvers.

outoftardis replied:
That was the input

PetrovKP replied:
And there is. We say that we support only one solver - SGD-minibatch for GPU. And we do not in any way support the L1 regulation of GPU.

outoftardis replied:
I can change this note to "L1 regularization is not supported on GPU." if it makes it clearer

averbukh commented on docs/source/dev_guide/algorithms/svm/support-vector-machine-classifier.rst:
```diff
@@ -43,14 +43,18 @@ belongs, the problem is to build a two-class Support Vector Machine
+
+|short_name| provides two methods to train the SVM model:
+
+- Boser method [Boser92]_
```
Need to add alternative names for clarification

outoftardis replied:
what are the alternative names?

averbukh replied:
I mean SMO (Sequential Minimal Optimization), but some addition clarification about difference between Boser's and Platt's algorithms. I suggest to consult with @PetrovKP about better formulation. He is ready to help with this.

PetrovKP replied:
I would suggest describing:

```
Boser method [Boser92] which is being decided the Sequential Minimal Optimization algorithm proposed Platt
```

outoftardis replied:
added description

averbukh commented on docs/source/dev_guide/includes/parameter_numeric_table.rst:
```diff
@@ -0,0 +1,21 @@
+.. * limitations under the License.
+.. *******************************************************************************/
+
+.. note::
+
+    This parameter can be an object of any class derived from ``NumericTable``
+    except ``PackedTriangularMatrix``, ``PackedSymmetricMatrix``, and
+    ``CSRNumericTable``.
```
Do we need dedicated file for that?

outoftardis replied:
these notes appear throughout the doc, it's easier to reference them each time that copy-paste

averbukh replied:
OK

outoftardis commented on docs/source/cpu-vs-gpu.rst:
```diff
@@ -91,7 +91,11 @@ Parameters
      - ``doShrinking`` is only supported for ``defaultDense`` method
    * - :ref:`dbscan`
-     - ``memorySavingMode`` could be ``false`` only on CPU
+     - 
```
removed it

averbukh replied:
Looks like it returned with latest commits. Please check

outoftardis replied:
there was an empty line, I removed it
the line 94 is not empty and is required to be there

outoftardis commented on docs/source/dev_guide/algorithms/k_nearest_neighbors/k-nearest-neighbors-knn-classifier.rst:
```diff
@@ -56,10 +53,21 @@ Given a positive integer parameter :math:`k` and a test observation
-[Patwary2016]
-that uses a space-partitioning data structure known as K-D tree. Each
-non-leaf node of a tree contains the identifier of a feature along
+The library provides kNN classification with the following methods:
```
could they be called search techniques?

averbukh replied:
As quick fix you can use use 'approaches' term, but actually it's about usage of the index for k-d tree case, and approach without any index for brute-force case. I suggest to consult author of those changes, because it's his responsibility to formulate it in the correct way.

outoftardis replied:
Ok, used "approaches" for now

outoftardis commented on docs/source/dev_guide/algorithms/k_nearest_neighbors/k-nearest-neighbors-knn-classifier.rst:
```diff
@@ -56,10 +53,21 @@ Given a positive integer parameter :math:`k` and a test observation
+K-D tree
+--------
+
+K-D tree, or a multidimensional binary search tree, is a space-partitioning data structure, where D is the dimension and K is
```
could you explain what does it mean in terms of documentation?

averbukh replied:
I see that linkage between knn and k-d tree was removed: "The library provides kNN classification based on multidimensional binary search tree (K-D tree, where D means the dimension and K means
the number of dimensions in the feature space)."

outoftardis replied:
fixed

outoftardis commented on docs/source/dev_guide/algorithms/k_nearest_neighbors/k-nearest-neighbors-knn-classifier.rst:
```diff
@@ -68,9 +76,19 @@ Feature vectors from any bucket belong to the region of the space
 Training Stage
 --------------
 
+Training using K-D Tree
```
Do you have a proposal how to change it?

averbukh replied:
You should ask the author of the LLD.

outoftardis replied:
I believe those changes were not reflected in LLD because the interfaces didn't change

averbukh replied:
What do you mean? Introduction of brute-force surely affected interface.

outoftardis commented on docs/source/dev_guide/algorithms/k_nearest_neighbors/k-nearest-neighbors-knn-classifier.rst:
```diff
@@ -99,12 +118,23 @@ threshold. Once the threshold is reached, PANDA stops growing the
+Training using Brute Force
+++++++++++++++++++++++++++
+
+During training with the Brute Force method, the algorithm copies all feature vectors from the training data set
```
@amgrigoriev could you help here?

amgrigoriev replied:
Brute force method always needs all training data. Thus, all data are stored in the  model. If data aren't stored, the method is unable to predict.

averbukh replied:
Storing - yes, copying - no. Just like for kd-trees

outoftardis replied:
fixed

outoftardis commented on docs/source/dev_guide/algorithms/k_nearest_neighbors/k-nearest-neighbors-knn-classifier.rst:
```diff
@@ -68,9 +76,19 @@ Feature vectors from any bucket belong to the region of the space
+Brute Force
+-----------
+
+The main idea of the Brute Force method is to calculate all of the distances from each query feature vector
+to each reference feature vector in the training data set and then select :math:`k` objects 
+that are closest to the query point.
```
what do you mean by "not calculated inside"?

averbukh replied:
We don't calculate distances in the implementation, while I see "to calculate all of the distances" in the doc. I confuses, because calculation of distances is really bad idea (but possbile) to do in the brute-force kNN.

To preserve consistency and common logic, I expect to to here explanation of the brute-force approach of kNN with reference to bibliography that can be used for further reading.

outoftardis replied:
@amgrigoriev  need your help here

amgrigoriev replied:
My suggestion: "Brute Force method calculates all of the squared distances from each query feature vector
to each reference feature vector in the training data set and then for every query feature vector selects :math:`k` objects from the training set that are closest to it."
In reality,  squared distances are incomplete ones but details can't be explained in a couple of sentences. It would be better to skip internal details in a brief description.

outoftardis commented on docs/source/dev_guide/algorithms/k_nearest_neighbors/k-nearest-neighbors-knn-classifier.rst:
```diff
@@ -56,10 +53,21 @@ Given a positive integer parameter :math:`k` and a test observation
-non-leaf node of a tree contains the identifier of a feature along
+The library provides kNN classification with the following methods:
+
+- K-D tree (only suppored on CPU)
+- Brute Force (only suppored on GPU)
```
would this explanation be sufficient?

```
kNN classification on CPU and GPU is based on different algorithms. 
On CPU, kNN classification uses K-D data structures to find nearest neighbors,
while on GPU the Brute Force search is used.
```

averbukh commented on docs/source/dev_guide/algorithms/dbscan/computation-batch.rst:
```diff
@@ -53,6 +53,9 @@ The DBSCAN clustering algorithm has the following parameters:
        It will require up to :math:`O(|\text{sum of sizes of all observations' neighborhoods}|)` of additional memory, 
        which in worst case can be :math:`O(|\text{number of observations}|^2)`. However, in general, performance may be better.
+
+       .. note:: ``memorySavingMode`` could be ``false`` only on CPU.
```
Unclear formulation. False is default value. Does it mean that algorithm will fail, if default value to be used? Anyway, 'could be' is bad for technical doc, some specific should be explained, like parameter to be ignored o GPU, or exception to be raised.

outoftardis replied:
@amgrigoriev could you clarify this?

amgrigoriev replied:
Algorithm will return error if memorySavingMode is true. It was a requirement of reviewers. Original behavior was to ignore the parameter. It is handled in python. My suggestion "is supported" instead of "could be".

outoftardis replied:
I propose this description:
* On GPU, the ``memorySavingMode`` flag can only be set to ``true``. You will get an error if the flag is set to ``false``.

amgrigoriev replied:
Agree

averbukh commented on docs/source/dev_guide/algorithms/dbscan/computation-batch.rst:
```diff
@@ -91,6 +94,10 @@ For more details, see :ref:`algorithms`.
 
+       .. note::
+
+         This parameter is not supported on GPU.
```
What it means for user? To be ignored? Exception to be raised?

outoftardis replied:
@amgrigoriev could you clarify?

amgrigoriev replied:
Weights are ignored

outoftardis replied:
fixed

averbukh commented on docs/source/dev_guide/algorithms/decision_forest/decision-forest.rst:
```diff
@@ -49,10 +49,22 @@ internal nodes :math:`t` represent a binary test (split) dividing their
+Inexact Histogram Computation Method
+++++++++++++++++++++++++++++++++++++
+
+In inexact histogram method only a selected subset of splits is considered for computation of a best split.
```
Can we add reference to bibliography to there details (how subset is specified and so on) can be found?

amgrigoriev replied:
There is no article the whole algorithm is based on exactly.

outoftardis replied:
@makart19 is there a reference we can add?

outoftardis replied:
no input

averbukh commented on docs/source/dev_guide/algorithms/k_nearest_neighbors/k-nearest-neighbors-knn-classifier.rst:
```diff
@@ -68,9 +76,19 @@ Feature vectors from any bucket belong to the region of the space
+Brute Force algorithm calculates the squared distances from each query feature vector
+to each reference feature vector in the training data set. Then,
+for each query feature vector it selects :math:`k` objects from the training set that are closest to that query feature vector.
+
```
Reference to bibliography is required.

outoftardis replied:
I think Alexey [answered to the wrong comment[(https://github.com/oneapi-src/oneDAL/pull/667#discussion_r448949701) that there is no article to refer to

averbukh replied:
Nobody said that it should be single reference and/or it should cover all aspects of our implementation. But absense of reference at all is not consistent.

outoftardis replied:
Added references

averbukh commented on docs/source/dev_guide/algorithms/k_nearest_neighbors/k-nearest-neighbors-knn-classifier.rst:
```diff
@@ -68,9 +76,19 @@ Feature vectors from any bucket belong to the region of the space
+Brute Force
+-----------
+
+Brute Force algorithm calculates the squared distances from each query feature vector
```
'Brute Force kNN algorithm' or 'Brute Force approach'.

outoftardis replied:
fixed

averbukh commented on docs/source/dev_guide/algorithms/optimization-solvers/solvers/stochastic-gradient-descent-algorithm.rst:
```diff
@@ -124,79 +124,83 @@ method parameter method:
+
+   * - ``batchIndices``
+     - ``defaultDense``, ``miniBatch``, ``momentum``
+     - ``NULL``
```
What NULL means here?

outoftardis replied:
This topic was converted from [classical DAAL doc](https://software.intel.com/content/www/us/en/develop/documentation/daal-programming-guide/top/algorithms/analysis/optimization-solvers/iterative-solver/stochastic-gradient-descent-algorithm/computation-9.html).  

Should it be "not applicable" instead of null?

outoftardis replied:
fixed

averbukh commented on docs/source/dev_guide/algorithms/svm/support-vector-machine-classifier.rst:
```diff
@@ -43,14 +43,18 @@ belongs, the problem is to build a two-class Support Vector Machine
+
+|short_name| provides two methods to train the SVM model:
+
+- Boser method [Boser92]_ that is the Sequential Minimal Optimization algorithm proposed by Platt [Platt98]_
```
@PetrovKP , add details about differences between Boser and Platt. According to your formulation, they are the same, but it isn't true. Boser in '92 proposed something different then Platt in '98.

PetrovKP replied:
```suggestion
- Boser method [Boser92]_ - performance-oriented variant of Boser [Boser92] and Platt [Platt98] algorithms.
```

PetrovKP commented on docs/source/dev_guide/algorithms/svm/support-vector-machine-classifier.rst:
```diff
@@ -201,50 +203,55 @@ Training and Prediction.
-   Batch Processing:
+    Batch Processing:
+    
+    -  :ref:`svm_two_class_dense_batch.cpp`
```
I renamed an example today https://github.com/oneapi-src/oneDAL/pull/696
```suggestion
    -  :ref:`svm_two_class_thunder_dense_batch.cpp`
```

outoftardis replied:
fixed

outoftardis commented:
@averbukh I've addressed your comments except the one about the bibliography reference, please take a look and say is there anything else that should be fixed?

outoftardis commented:
@averbukh do you have more comments or is this ok to merge?

outoftardis commented:
@averbukh I've added the references for kNN
